### ISSUE TITLE
HubConnection.DisposeAsync causes deadlock if it was called synchronously

### ISSUE THREAD

**Issue by DenSmoke** [dotnet/aspnetcore#14695]
HubConnection.DisposeAsync causes deadlock if it was called synchronously 

1. Use `Microsoft.AspNetCore.SignalR.Client 3.0.0` in WPF .NET Core 3.0 app
2. Code (`HubConnection` should be started before `Current.Exit` is invoked:
```C#
class App : Application
{
   public HubConnection HubConnection {get; private set; }
   public App()
   {
      Current.Exit += (s, e) => HubConnection.DisposeAsync().Wait();
      InitializeComponent();
   }
}
```

As I see in source code of HubConnection `ConfigureAwait(false)` is never used.

Is any workaround for this case?
Currently I solved this by `_ = HubConnection.DisposeAsync()` but this is not good solution.

P.S. For some reason I can't reproduce it in xUnit test - deadlock doesn't occur

**Comment by BrennanConroy** [comment]
This has nothing to do with `ConfigureAwait(false)`, `HubConnection` uses a fun little trick that lets us not put `ConfigureAwait` everywhere in our code. https://github.com/aspnet/AspNetCore/blob/fd060ce8c36ffe195b9e9a69a1bbd8fb53cc6d7c/src/SignalR/common/Shared/ForceAsyncAwaiter.cs#L9

----

If you put `Task.Delay(1000).ConfigureAwait(false).GetAwaiter().GetResult();` in your `Current.Exit` handler you'll see that it also causes a deadlock.

I don't know the details of what's going on internally in WPF but it's generally a bad idea to call `.Wait()` in UI frameworks.

Maybe you can ask how to do async disposal on the [WPF repo](https://github.com/dotnet/wpf/issues)?

**Comment by davidfowl** [comment]
You can't do .Wait in UI applications, this is a very well known issue:
- https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html
- https://devblogs.microsoft.com/pfxteam/await-and-ui-and-deadlocks-oh-my/

As much as I hate to recommand it, async void is supported in UI based applications (or you can fire and forget).

**Comment by DenSmoke** [comment]
I know that `IAsyncDisposable` is the future and all that, but it would be nice to have simple `Dispose()` tbh. I also use `HubConnection` in my own services implementing only `IDisposable` and I forced to write `.Wait()` , use discard or `async void` and then some people realize that it doesn't work properly in their apps.

**Comment by davidfowl** [comment]
We don't plan to add any synchronous APIs to the `HubConnection`. It's waiting on something asynchronous to finish and is inherently asynchronous. The fact that you implement these in services that implement `IDisposable` is itself a problem as it just tries to ignore the issue.

**Comment by DenSmoke** [comment]
Hmm, it is very weird decision. Some APIs in .NET Core implement both IDisposable and IAsyncDisposable. I think it is good practice and no inconsistency occurs

**Comment by anurse** [comment]
A SignalR connection **cannot** be disposed synchronously because of the network I/O involved. Adding a `Dispose` API would mislead users. Some APIs have the ability to be disposed both sychronously and asynchronously, and in other cases (like `System.IO.Stream`) we had to support both since it's an abstract base class.

How to react depends on what you're intending to do here. Blocking the shutdown on an async process is quite difficult here since I don't believe WPF gives you an async way to do so. If we provided a synchronous `IDisposable` it would likely deadlock the same way, since we'd have to just use `.Wait()`. Hiding this kind of behavior seems like it would confuse users more.

**Comment by DenSmoke** [comment]
FYI, workaround for this issue:

```C#
class App : Application
{
    public HubConnection HubConnection { get; private set; }

    public App()
    {
        Current.Exit += (s, e) => 
            Task.Run(async() => await HubConnection
                    .DisposeAsync()
                    .ConfigureAwait(false))
                .Wait();
        InitializeComponent();
    }
}
```
